### DISCUSSION TITLE
Early stop saving best checkpoints

### DISCUSSION THREAD

**Discussion by minwang-ai** [Lightning-AI/pytorch-lightning#10399] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule] [answered]
Hi all, do you know how to save the best model? Since pytorchlighting 's earlystop callback will monitor val_loss and if val_loss stop decreasing, it will stop training automaticlly. In this case, the checkpoint of the final model would be the final epoch (the val_loss starts to increase). Can I save epoch 5 or 6 (before val_loss increasing) as the best model?
![CC51157F-0EBE-4D35-B343-5C85F17A4EC2](https://user-images.githubusercontent.com/75472853/140654686-02726372-16fd-40b4-9d7d-ee323ae500ba.jpeg)

**Comment by rohitgr7** [comment, marked as answer]
you can specify
```py
ModelCheckpoint(monitor='val_loss', save_top_k=1, ...)
```
this will save the best model.

**Reply by minwang-ai** [reply]
Thank you!
Should I set every_n_epochs=1 right?

For example:
```python
checkpoint_callback = ModelCheckpoint(
        monitor="val_loss",
        dirpath="/home/mwa7rng/mwa7rng/output/FFB6D/checkpoints/",
        filename="{epoch}-{val_loss:.2f}-{val_iou:.2f}",
        save_top_k=1,        # save the best model
        mode="min",
        every_n_epochs=1
    )
```

**Reply by rohitgr7** [reply]
yes

**Reply by minwang-ai** [reply]
After using checkpoints callback, I found that the checkpoints of all my experiments are saved in the dirpath. Before that, the checkpoints of each run are saved in the subfolder of tensorboard log folder, i.e., logs/exp_name/exp-version/checkpoints. **How can I save checkpoints with exp_name when I use callback?**

In the docs, it shows:  _By default, dirpath is None and will be set at runtime to the location specified by Trainer’s default_root_dir or weights_save_path arguments, and **if the Trainer uses a logger, the path will also contain logger name and version.**_

However, it does not contain logger name and version when I use checkpoint callback. Different experiments’s checkpoints are saved at same path. 
![EC605107-7F04-4054-9122-66539E35B9B1](https://user-images.githubusercontent.com/75472853/140901579-d8806335-f9ce-4a0c-adf7-3a7017015968.jpeg)

**Reply by rohitgr7** [reply]
can you try setting `ModelCheckpoint(dirpath=None, ...)`?

**Reply by minwang-ai** [reply]
> can you try setting `ModelCheckpoint(dirpath=None, ...)`?

Thank you ! The problem solved! (It would be better to update docs😊)

**Reply by rohitgr7** [reply]
I think it's already updated:
https://github.com/PyTorchLightning/pytorch-lightning/blob/09cf167237e867f1ec67a5db87e5a02c2cea4b69/pytorch_lightning/callbacks/model_checkpoint.py#L65-L69
